insonmnia/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,351 @@
+package gatekeeper
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/noxiouz/zapctx/ctxlog"
+	"github.com/sonm-io/core/blockchain"
+	"github.com/sonm-io/core/util"
+	"go.uber.org/zap"
+)
+
+type Gatekeeper struct {
+	key *ecdsa.PrivateKey
+	cfg *Config
+
+	logger *zap.Logger
+
+	bch blockchain.API
+
+	in  blockchain.SimpleGatekeeperAPI
+	out blockchain.SimpleGatekeeperAPI
+
+	freezingTime time.Duration
+
+	spentToday *big.Int
+	lastDay    *time.Time
+
+	mu sync.Mutex
+}
+
+func NewGatekeeper(ctx context.Context, key *ecdsa.PrivateKey, cfg *Config) (*Gatekeeper, error) {
+	logger := ctxlog.GetLogger(ctx)
+
+	bch, err := blockchain.NewAPI(blockchain.WithConfig(cfg.Blockchain))
+	if err != nil {
+		return nil, err
+	}
+
+	var in blockchain.SimpleGatekeeperAPI
+	var out blockchain.SimpleGatekeeperAPI
+
+	if cfg.Gatekeeper.Direction == "masterchain" {
+		in = bch.SidechainGate()
+		out = bch.MasterchainGate()
+	} else if cfg.Gatekeeper.Direction == "sidechain" {
+		in = bch.MasterchainGate()
+		out = bch.SidechainGate()
+	}
+
+	keeper, err := out.GetKeeper(ctx, crypto.PubkeyToAddress(key.PublicKey))
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("start gatekeeper instance",
+		zap.String("direction", cfg.Gatekeeper.Direction),
+		zap.Duration("delay", cfg.Gatekeeper.Delay),
+		zap.String("key", crypto.PubkeyToAddress(key.PublicKey).String()),
+		zap.String("day limit", keeper.DayLimit.String()),
+		zap.String("spent today", keeper.SpentToday.String()))
+
+	if keeper.DayLimit.Cmp(big.NewInt(0)) == 0 {
+		return nil, fmt.Errorf("used key is not keeper")
+	}
+
+	if keeper.Frozen {
+		return nil, fmt.Errorf("keeper with given key is frozen")
```

Comment:
you're logging "start gatekeeper" and then performs some unconditionally required pre-flight checks. Move this above.

Reply:
fixed